config/sql/config.go

```diff
@@ -93,4 +83,19 @@ func Configure(p *config.Provider) {
 		)
 		r.UseAsync = true
 	})
+	p.AddResourceConfigurator("azurerm_mssql_server_transparent_data_encryption", func(r *config.Resource) {
+		r.Version = common.VersionV1Alpha2
+		r.References = config.References{
+			"server_id": config.Reference{
+				Type:      "MSSQLServer",
+				Extractor: rconfig.ExtractResourceIDFuncPath,
+			},
+			"key_vault_key_id": config.Reference{
+				Type:      rconfig.VaultKeyReferencePath,
+				Extractor: rconfig.ExtractResourceIDFuncPath,
+			},
+		}
+		r.ExternalName = config.IdentifierFromProvider
```

Comment:
Hi @ytsarev,
We normally use `config.IdentifierFromProvider` for resources for which the external name has a Cloud-provider assigned component. The classic example is the AWS VPC resource, which is assigned a generated ID by AWS. Looking at the output of your tests in the PR description, it looks like that we have the following as an example for the external name of an `MSSQLServerTransparentDataEncryption`:
`/subscriptions/<subcription>/resourceGroups/<rg name>/providers/Microsoft.Sql/servers/<server name>/encryptionProtector/current`

It looks like this external name does not have any Cloud-provider assigned part. If this is really the case, like we do in other `provider-jet-azure` resources, we can consider using `NameAsIdentifier` but with overrides.

Reply:
Btw, if I understand correctly, ID for a `MSSQLServerTransparentDataEncryption` resource is: `<ID for the parent MSSQL server>/encryptionProtector/current`, is this correct? It looks like there is no identifying info specific to the data encryption instance. If there can be no multiple of such configurations, I think it makes sense but asking just to make sure.

Reply:
@ulucinar you are right, it does not have its own unique id, only the reference of associated mssql server, see https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mssql_server_transparent_data_encryption#import

Reply:
@ulucinar just to double sure, we don't need to switch to `NameAsIdentifier` because `current` is actually not a custom name, correct?

Reply:
Hi @ytsarev,
I was suggesting something in the lines of:
```suggestion
		r.ExternalName = config.NameAsIdentifier
		r.ExternalName.GetIDFn = func(_ context.Context, _ string, parameters map[string]interface{}, _ map[string]interface{}) (string, error) {
			return fmt.Sprintf("%s/encryptionProtector/current", parameters["server_id"]), nil
		}
```

although I did not verify it actually works.

Reply:
Thanks @ulucinar , I am definitely not against it, but just trying to understand what is the motivation to manually construct the identifier when it is already correctly returned from the provider as visible in the test?
```
transparent-encryption-example   True    True     /subscriptions/038f2b7c-3265-43b8-8624-c9ad5da610a8/resourceGroups/mssql-server-transparent-data-encryption-example/providers/Microsoft.Sql/servers/mssql-server-transparent-data-encryption-example/encryptionProtector/current   33m

```

Is it performance/semantics or both?

Reply:
Thank you for bearing with me Yury. Initially, we were using Azure fully qualified IDs as the external names of all of the `provider-jet-azure` resources. Then support for resource external-name configurations were introduced ([`config.ExternalName`](https://github.com/crossplane/terrajet/blob/6142eefbcf5b4e6490a4a39f8ceffa8810911d5f/pkg/config/resource.go#L76)). Then we switched to [`config.NameAsIdentifier`](https://github.com/crossplane/terrajet/blob/6142eefbcf5b4e6490a4a39f8ceffa8810911d5f/pkg/config/defaults.go#L31) for managed resources for which Azure does not assign a random ID component as we've discussed before. But when you use `config.NameAsIdentifier` as the external-name configuration, you need to override the `config.ExternalName.GetIDFn` as in `config.NameAsIdentifier`'s implementation, it returns the value of the external-name annotation as the ID (and which is in turn initialized from `metadata.name`), whereas Azure expects its fully-qualified IDs.

Thus, using Azure IDs as managed resource external-names requires less manual work with our implementation but back then we had discussions involving the Crossplane resource model (XRM, https://github.com/crossplane/crossplane/blob/master/design/one-pager-managed-resource-api-design.md) and we opted for the _normalization_ of the external-names of the `provider-jet-azure` resources. So, the subscription ID is already specified as part of the referenced `ProviderConfig`. Similarly, the resource group under which the external resource (infra resource) lives is already specified as part of the resource spec (e.g., `spec.forProvider.resourceGroupName`). In this specific case of `MSSQLServerTransparentDataEncryption`, the parent MSSQL server's identifying info should already be available in `spec.forProvider.serverID`). Because these ID components are already supplied as input, we did not want them to be part of the external-name. Please also note that when importing resources, users of the managed resource APIs may need to specify the external-name annotation unless we are using `metadata.name` as the external-name.

Thus, the answer to your question is normalization and the associated improved UX I would say.

Reply:
We can also mention API-consistency as a further reason to explore using `config.NameAsIdentifier` for this resource. Given that we are using `metadata.name` as the external-names of most Azure resources (for those no Azure-assigned random component exists), it will be more consistent with the other APIs.

Reply:
@ulucinar thanks a lot for the detailed explanation, I really appreciate it!

I've tried `r.ExternalName = config.NameAsIdentifier` but if failed with `: Extraneous JSON object property: No argument or block type is named \"name\"`

```
1.648590066015713e+09	DEBUG	events	Warning	{"object": {"kind":"MSSQLServerTransparentDataEncryption","name":"transparent-encryption-example","uid":"20c551f7-d430-46a5-8e4c-c5cf28d1cdf0","apiVersion":"sql.azure.jet.crossplane.io/v1alpha2","resourceVersion":"191645"}, "reason": "CannotObserveExternalResource", "message": "cannot run refresh: refresh failed: Extraneous JSON object property: No argument or block type is named \"name\".: File name: main.tf.json"}
```
Apparently that's because this specific resource does not have a `name` parameter. What is our strategy in this case?

Reply:
Hi @ytsarev,
Could you please also set the following?
```go
r.ExternalName.SetIdentifierArgumentFn = config.NopSetIdentifierArgument
```